src/Plugin/resource/CsrfToken.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\restful\Plugin\resource\CsrfToken.
+ */
+
+namespace Drupal\restful\Plugin\resource;
+use Drupal\restful\Resource\ResourceManager;
+
+/**
+ * Class CsrfToken
+ * @package Drupal\restful\Plugin\resource
+ *
+ * @Resource(
+ *   name = "csrf_token:1.0",
+ *   resource = "csrf_token",
+ *   label = "CSRF Token",
+ *   description = "Resource that provides CSRF Tokens.",
+ *   authenticationTypes = TRUE,
+ *   authenticationOptional = TRUE,
+ *   majorVersion = 1,
+ *   minorVersion = 0
+ * )
+ */
+class CsrfToken extends Resource implements ResourceInterface {
+
+  /**
+   * Constructs a CsrfToken object.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    // Set dynamic options that cannot be set in the annotation.
+    $plugin_definition = $this->getPluginDefinition();
+    $plugin_definition['menuItem'] = variable_get('restful_hook_menu_base_path', 'api') . '/session/token';
+
+    // Store the plugin definition.
+    $this->pluginDefinition = $plugin_definition;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function publicFields() {
+    return array(
+      'X-CSRF-Token' => array(
+        'callback' => '\Drupal\restful\Plugin\resource\CsrfToken::getCsrfToken',
+      ),
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function dataProviderFactory() {
+    return new DataProvider\DataProviderNull($this->getRequest(), $this->getFieldDefinitions(), $this->getAccount(), array());
```

Comment:
If I understand correctly we use `DataProviderNull` because we calculate this on the fly. If so maybe worth adding a comment.

Second thing is that it's very verbose considering we just want a null implementation. Maybe we can change the signature of  `\Drupal\restful\Plugin\resource\DataProvider\DataProvider::__construct(RequestInterface $request = UNLL, ResourceFieldCollectionInterface $field_definitions = NULL, $account = NULL, array $options = NULL, $langcode = NULL);``

So here we could just `return new DataProvider\DataProviderNull();`